1811 - elaborate on README.md in packages (#1818)
defenseunicorns/zarf: packages/README.md

**Packages**
- [distros](#distros)
- [gitea](#gitea)
- [logging-pgl](#logging-pgl)
- [zarf-agent](#zarf-agent)
- [zarf-registry](#zarf-registry)

### Distros

The distros package adds optional capabilities for spinning up and tearing down clusters.  Currently, the following distros are supported:

- [EKS](https://aws.amazon.com/eks/) - Zarf deploys and tears down using the `eksctl` binary under the hood. See how it's done in the EKS package's [`zarf.yaml`](./distros/eks/zarf.yaml) and checkout the [EKS package's config](./distros/eks/eks.yaml) for more information.

- [k3s](https://k3s.io/) - Zarf deploys and tears down using the `k3s` service under the hood. See how it's done in the k3s package's [`zarf.yaml`](./distros/k3s/common/zarf.yaml).

#### Usage Examples  
  
**EKS**  - Create/Deploy EKS cluster. (requires `eksctl` credentials)

```bash
zarf package create packages/distros/eks -o build --confirm

zarf package deploy build/zarf-package-distro-eks-multi-x.x.x.tar.zst --components=deploy-eks-cluster --set=CLUSTER_NAME='zarf-nightly-eks-e2e-test',INSTANCE_TYPE='t3.medium' --confirm
```

See the [nightly-eks test](../.github/workflows/nightly-eks.yml) for another example.

> **k3s** - Create/Deploy a k3s cluster.

Reviewer: since we listed a pre-req for EKS, we should specify that this only works when you're running as a root user on a linux machine here

Author: @YrrepNoj  So it's `k3s`, correct? LMK if you want to see that note moved into the examples if it is good there.

Author: ah yes, definitely agree. Added more uniformity there:

- [Line 25](https://github.com/cmwylie19/zarf/blob/d9f335eb018ab9a42d0234547f29b2e509d6fcf6/packages/README.md?plain=1#L25)

- [Line 37](https://github.com/cmwylie19/zarf/blob/d9f335eb018ab9a42d0234547f29b2e509d6fcf6/packages/README.md?plain=1#L37)

- [zarf-registry](#zarf-registry)

### Distros

The distros package adds optional capabilities for spinning up and tearing down clusters.  Currently, the following distros are supported:

- [EKS](https://aws.amazon.com/eks/) - Zarf deploys and tears down using the `eksctl` binary under the hood. See how it's done in the EKS package's [`zarf.yaml`](./distros/eks/zarf.yaml) and checkout the [EKS package's config](./distros/eks/eks.yaml) for more information.

- [k3s](https://k3s.io/) - Zarf deploys and tears down using the `k3s` service under the hood. See how it's done in the k3s package's [`zarf.yaml`](./distros/k3s/common/zarf.yaml).

#### Usage Examples  
  
**EKS**  - Create/Deploy EKS cluster. (requires `eksctl` credentials)

```bash
zarf package create packages/distros/eks -o build --confirm

zarf package deploy build/zarf-package-distro-eks-multi-x.x.x.tar.zst --components=deploy-eks-cluster --set=CLUSTER_NAME='zarf-nightly-eks-e2e-test',INSTANCE_TYPE='t3.medium' --confirm
```

See the [nightly-eks test](../.github/workflows/nightly-eks.yml) for another example.

**k3s** - Create/Deploy a k3s cluster.

```bash
# Linux with `systemd` and `root` access only (no `sudo`)

> zarf init --components=git-server

Reviewer 2: ```suggestion
zarf init --components=k3s
```

did you mean to use `k3s` here?

Author: ah yes, i definitely did. Good eye!

Reviewer 3: My bad :sweat_smile:

```

### Gitea

> Users who rely heavily on GitOps find it useful to deploy an internal Git repository.  Zarf uses [Gitea](https://gitea.io/en-us/) to provide this functionality.  The Gitea package deploys a Gitea instance to the cluster and configures it to use the credentials in the `private-git-server` secret in the zarf namespace.

Reviewer 2: ```suggestion
Users who rely heavily on GitOps find it useful to deploy an internal Git repository.  Zarf uses [Gitea](https://gitea.io/en-us/) to provide this functionality.  The Gitea package deploys a Gitea instance to the cluster and configures it to use the credentials in the `private-git-server` secret in the Zarf namespace.
```

Use caps when referring to Zarf proper, usage in `code` or other such cases can be lowercase.

Reviewer 2: https://docs.zarf.dev/docs/contribute-to-zarf/style-guide#introduction

Author: giving this a read now. TY

_usage_

```bash
zarf init --components=git-server
```

### Logging PGL

> The Logging PGL package deploys the Promtail Grafana & Loki stack which aggreates logs from different containers and presents them in a web dashboard.  This is useful as a quick way to get logging into a cluster when you otherwise wouldn't be bringing over a logging stack.

Reviewer 2: ```suggestion
The Logging PGL package deploys the Promtail, Grafana, and Loki stack which aggregates logs from different containers and presents them in a web dashboard.  This is useful as a quick way to get logging into a cluster when you otherwise wouldn't be bringing over a logging stack.
```

Author: TY

_usage_

```bash
zarf init --components=logging
```

### Zarf Agent

> The Zarf Agent is a mutating admission controller used to modify the image property within the PodSpec. The purpose is to redirect it to Zarf's configured registry instead of the the original registry (such as DockerHub, GCR, or Quay). Additionally, the webhook attaches the appropriate `ImagePullSecret` for the seed registry to the pod. This configuration allows the pod to successfully retrieve the image from the seed registry, even when operating in an air-gapped environment.

Reviewer 2: ```suggestion
The Zarf Agent is a mutating admission controller used to modify the image property within a PodSpec. The purpose is to redirect it to Zarf's configured registry instead of the the original registry (such as DockerHub, GHCR, or Quay). Additionally, the webhook attaches the appropriate `ImagePullSecret` for the seed registry to the pod. This configuration allows the pod to successfully retrieve the image from the seed registry, even when operating in an air-gapped environment.
```

```bash

> kubectl get deploy -n zarf agent-hook 
> ```
> 
> output
> 
> ```bash
> NAME         READY   UP-TO-DATE   AVAILABLE   AGE
> agent-hook   2/2     2            2           17m

Reviewer 2: ```suggestion
$ kubectl get deploy -n zarf agent-hook 

NAME         READY   UP-TO-DATE   AVAILABLE   AGE
agent-hook   2/2     2            2           17m
```

my preferred method for showing the output of a command is to add `$` to the beginning of each command and show the command + output in one codeblock

Reviewer 2: additionally, Zarf vendors `kubectl`, so this command _could_ be `zarf tools kubectl ...`

Author: like it!

```

See the code:
- [zarf-agent](../src/internal/agent)
- [transform-library](../src/pkg/transform)

### Zarf Registry

> The Zarf internal registry is utilized to store container images for use in air-gapped environments.  The registry is deployed as a `Deployment` with a single replica and  a `PersistentVolumeClaim` to store the images.  Credentials for basic authentication are autogenerated and stored within a secret in the `zarf` namespace.

Reviewer 2: Would add a note that the internal registry is also HTTP only.

Author: sounds good - done.

- [logging-pgl](#logging-pgl)
- [zarf-agent](#zarf-agent)
- [zarf-registry](#zarf-registry)

### Distros

The distros package adds optional capabilities for spinning up and tearing down clusters.  Currently, the following distros are supported:

- [EKS](https://aws.amazon.com/eks/) - Zarf deploys and tears down using the `eksctl` binary under the hood. See how it's done in the EKS package's [`zarf.yaml`](./distros/eks/zarf.yaml) and checkout the [EKS package's config](./distros/eks/eks.yaml) for more information.

- [k3s](https://k3s.io/) - Zarf deploys and tears down using the `k3s` service under the hood. See how it's done in the k3s package's [`zarf.yaml`](./distros/k3s/common/zarf.yaml).

#### Usage Examples  
  
**EKS**  - Create/Deploy EKS cluster. (requires `eksctl` credentials)

```bash
zarf package create packages/distros/eks -o build --confirm

zarf package deploy build/zarf-package-distro-eks-multi-x.x.x.tar.zst --components=deploy-eks-cluster --set=CLUSTER_NAME='zarf-nightly-eks-e2e-test',INSTANCE_TYPE='t3.medium' --confirm
```

See the [nightly-eks test](../.github/workflows/nightly-eks.yml) for another example.

**k3s** - Create/Deploy a k3s cluster.

```bash

> zarf package create packages/distros/k3s -o build --confirm
> 
> zarf package deploy build/zarf-init-arm64-vx.x.x.tar.zst --confirm

Reviewer 3: ```suggestion
# Linux with `systemd` and `root` access only (no `sudo`)
zarf init --components=git-server
```

Since `k3s` is included in the default init package we could just show them this as a simpler way for now.

```

### Gitea

Users who rely heavily on GitOps find it useful to deploy an internal Git repository.  Zarf uses [Gitea](https://gitea.io/en-us/) to provide this functionality.  The Gitea package deploys a Gitea instance to the cluster and configures it to use the credentials in the `private-git-server` secret in the zarf namespace.

_usage_

```bash
zarf init --components=git-server
```

### Logging PGL

> The Logging PGL package deploys the Promtail Grafana & Loki stack which aggreates logs from different containers and presents them in a web dashboard.

Reviewer 3: ```suggestion
The Logging PGL package deploys the Promtail Grafana & Loki stack which aggreates logs from different containers and presents them in a web dashboard.  This is useful as a quick way to get logging into a cluster when you otherwise wouldn't be bringing over a logging stack.
```
A quick why you would want this may be useful

_usage_

```bash
zarf init --components=logging
```

### Zarf Agent

> The Zarf agent is a mutating admission controller used to modify the image property within the PodSpec. The purpose is to redirect it to Zarf's configured registry instead of the the original registry (such as DockerHub, GCR, or Quay). Additionally, the webhook attaches the appropriate `ImagePullSecret` for the seed registry to the pod. This configuration allows the pod to successfully retrieve the image from the seed registry, even when operating in an air-gapped environment.

Reviewer 3: ```suggestion
The Zarf Agent is a mutating admission controller used to modify the image property within the PodSpec. The purpose is to redirect it to Zarf's configured registry instead of the the original registry (such as DockerHub, GCR, or Quay). Additionally, the webhook attaches the appropriate `ImagePullSecret` for the seed registry to the pod. This configuration allows the pod to successfully retrieve the image from the seed registry, even when operating in an air-gapped environment.
```

Small thing but most of the other times we say Agent we capitalize it

```bash
kubectl get deploy -n zarf agent-hook 
```

output

```bash
NAME         READY   UP-TO-DATE   AVAILABLE   AGE
agent-hook   2/2     2            2           17m
```

See the code:
- [zarf-agent](../src/internal/agent)
- [transform-library](../src/pkg/transform)

### Zarf Registry

> The Zarf internal registry is utilized to store container images for use in air-gapped environments.  The registry is deployed as a deployment with a single replica.  The registry is configured to use a persistent volume claim to store the images. The registry is configured to use basic authentication.  The username and password are stored in a secret in the zarf namespace.

Reviewer 3: ```suggestion
The Zarf internal registry is utilized to store container images for use in air-gapped environments.  The registry is deployed as a `Deployment` with a single replica and  a `PersistentVolumeClaim` to store the images.  Credentials for basic authentication are autogenerated and stored within a secret in the `zarf` namespace.
```
Stylistic, but the flow could be a bit better on this one - I also like adding `backticks` to things to visually breakup sentences
